storage/series.go

```diff
@@ -413,3 +446,118 @@ func ExpandChunks(iter chunks.Iterator) ([]chunks.Meta, error) {
 	}
 	return result, iter.Err()
 }
+
+// RecodingAppender is a tsdb.Appender that recodes histogram samples if needed during appends.
+// It takes an existing appender and a chunk to which samples are appended.
+type RecodingAppender struct {
+	chk *chunkenc.Chunk
+	app chunkenc.Appender
+}
+
+func NewRecodingAppender(chk *chunkenc.Chunk, app chunkenc.Appender) *RecodingAppender {
+	return &RecodingAppender{
+		chk: chk,
+		app: app,
+	}
+}
+
+// Append appends a float sample to the appender.
+func (a *RecodingAppender) Append(t int64, v float64) {
+	a.app.Append(t, v)
+}
+
+// AppendHistogram appends a histogram sample to the underlying chunk.
+// It returns false if the sample cannot be appended.
+func (a *RecodingAppender) AppendHistogram(t int64, h *histogram.Histogram) bool {
+	app, _ := a.app.(*chunkenc.HistogramAppender)
+
+	if app.NumSamples() > 0 {
+		var (
+			pForwardInserts, nForwardInserts   []chunkenc.Insert
+			pBackwardInserts, nBackwardInserts []chunkenc.Insert
+			pMergedSpans, nMergedSpans         []histogram.Span
+			okToAppend                         bool
+		)
+		switch h.CounterResetHint {
+		case histogram.GaugeType:
+			if app != nil {
```

Comment:
The `nil` is of no use anymore here and below and in AppendFloatHistogram because you are calling `app.NumSamples()` above.